### PR TITLE
OpenTelemetry Integration

### PR DESCRIPTION
**Pull request by Gazza7205**
## OpenTelemetry Integration
 - Added OpenTelemetry integration, limited to metrics
 - Example with Layer7 Operator Dashboard using the Grafana LGTM stack

## Gateway Updates
 - Updated Gateway and Repository controllers
 - Added NodePort port config for Gateway Services

### CODE REVIEW

File: internal/controller/gateway/controller.go [hunk 1]
```diff
@@ -67,85 +103,230 @@ func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
 
-	err = reconcile.Route(ctx, params)
+	reconcileLatency, err := meter.Float64Histogram(otelMetricPrefix+"operator_gw_reconciler_latency",
+		metric.WithDescription("gateway controller reconcile latency"), metric.WithUnit("ms"))
 	if err != nil {
-		return ctrl.Result{}, err
+		return err
 	}
 
-	err = reconcile.HorizontalPodAutoscaler(ctx, params)
```
Comment (APIM-KiranVaddadi) [inline internal/controller/gateway/controller.go:103]: As its called in look need to have type attribute (operation) for the counters.

File: internal/controller/repository/controller.go [hunk 2]
```diff
@@ -81,5 +98,55 @@ func (r *RepositoryReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+	reconcileLatency, err := meter.Float64Histogram(otelMetricPrefix+"operator_repo_reconciler_latency",
+		metric.WithDescription("repository controller reconcile latency"), metric.WithUnit("ms"))
+	if err != nil {
+		return err
+	}
+
+	duration := time.Since(start)
+	reconcileLatency.Record(ctx, duration.Seconds(),
+		metric.WithAttributes(
+			attribute.String("pod", hostname),
```
[the remark this answers is not in the document]
Comment (Gazza7205) [reply internal/controller/repository/controller.go:147]: This is intended to cover the reconcile loop only. Additional fields are captured in the scheduled sync task.

File: pkg/gateway/reconcile/metrics.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	"strings"
+	"time"
+
+	"github.com/caapim/layer7-operator/pkg/util"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+func captureGraphmanMetrics(ctx context.Context, params Params, start time.Time, podName string, bundleType string, bundleName string, sha1sum string, hasError bool) error {
```
[the remark this answers is not in the document]
Comment (Gazza7205) [reply pkg/gateway/reconcile/metrics.go:14]: repo name and commitId are captured - this is not limited to repositories.

File: internal/controller/gateway/controller.go [hunk 4]
```diff
@@ -67,85 +103,230 @@ func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
+	reconcileLatency.Record(ctx, duration.Seconds(),
+		metric.WithAttributes(
+			attribute.String("pod", hostname),
+			attribute.String("namespace", operatorNamespace),
+		))
 
-func (r *GatewayReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	gatewayRepoRefs.Record(ctx, int64(len(gateway.Spec.App.RepositoryReferences)),
+		metric.WithAttributes(
+			attribute.String("pod", hostname),
```
[the remark this answers is not in the document]
Comment (Gazza7205) [reply internal/controller/gateway/controller.go]: Updated

File: pkg/repository/reconcile/repository.go [hunk 5]
```diff
@@ -182,11 +189,12 @@ func syncRepository(ctx context.Context, params Params) error {
 		repository.Status = repoStatus
 		err = params.Client.Status().Update(ctx, &repository)
 		if err != nil {
+			_ = captureRepositorySyncMetrics(ctx, params, start, commit, true)
 			params.Log.Info("failed to update repository status", "namespace", repository.Namespace, "name", repository.Name, "error", err.Error())
 		}
 		params.Log.Info("reconciled", "name", repository.Name, "namespace", repository.Namespace, "commit", commit)
 	}
-
+	_ = captureRepositorySyncMetrics(ctx, params, start, commit, false)
```
[the remark this answers is not in the document]
Comment (Gazza7205) [reply pkg/repository/reconcile/repository.go:198]: updated

File: internal/controller/gateway/controller.go [hunk 6]
```diff
@@ -67,85 +103,230 @@ func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
+			_ = captureMetrics(ctx, params, start, true)
 			return ctrl.Result{}, err
 		}
-		return ctrl.Result{}, err
 	}
 
-	err = reconcile.Secrets(ctx, params)
-	if err != nil {
-		return ctrl.Result{}, err
+	_ = captureMetrics(ctx, params, start, false)
```
[the remark this answers is not in the document]
Comment (Gazza7205) [reply internal/controller/gateway/controller.go]: Capturing reconcile as a whole vs. every operation.

File: internal/controller/gateway/controller.go [hunk 7]
```diff
@@ -67,85 +103,230 @@ func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
 
-	err = reconcile.GatewayStatus(ctx, params)
+	gatewayReconcileFailure, err := meter.Int64Counter(otelMetricPrefix+"operator_gw_reconcile_failure",
+		metric.WithDescription("gateway reconcile failure"))
 	if err != nil {
-		return ctrl.Result{}, err
+		return err
 	}
 
-	err = reconcile.ConfigMaps(ctx, params)
```
[the remark this answers is not in the document]
Comment (Gazza7205) [reply internal/controller/gateway/controller.go:118]: Updated

File: internal/controller/gateway/controller.go (hunk 1 above)
Comment (Gazza7205) [reply internal/controller/gateway/controller.go:103]: Updated

**Review by APIM-KiranVaddadi — APPROVED** [review]

**Pull request merged**